<?xml version="1.0"?>
<robot>
    <gazebo reference="arm_base_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>
    <gazebo reference="link_1_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>
    <gazebo reference="link_2_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>
    <gazebo reference="link_3_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>
    <gazebo reference="gripper_base_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>
    <gazebo reference="gripper_finger_left_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>
    <gazebo reference="gripper_finger_right_joint">
        <implicitSpringDamper>true</implicitSpringDamper>
    </gazebo>

    <gazebo reference="front_left_wheel_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>500000.0</kp>
        <kd>10.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>0.1</maxVel>
        <fdir1>1 0 0</fdir1>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>

    <gazebo reference="front_right_wheel_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>500000.0</kp>
        <kd>10.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>0.1</maxVel>
        <fdir1>1 0 0</fdir1>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>

    <gazebo reference="back_left_wheel_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>500000.0</kp>
        <kd>10.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>0.1</maxVel>
        <fdir1>1 0 0</fdir1>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>

    <gazebo reference="back_right_wheel_link">
        <mu1>0.1</mu1>
        <mu2>0.1</mu2>
        <kp>500000.0</kp>
        <kd>10.0</kd>
        <minDepth>0.001</minDepth>
        <maxVel>0.1</maxVel>
        <fdir1>1 0 0</fdir1>
        <!-- <material>Gazebo/FlatBlack</material> -->
    </gazebo>


    <gazebo>
        <plugin name="mobile_base_controller" filename="libgazebo_ros_diff_drive.so">
            <ros>
                <namespace>/demo</namespace>

                <remapping>cmd_vel:=cmd_demo</remapping>
                <remapping>odom:=odom_demo</remapping>
            </ros>

            <!-- Update rate -->
            <update_rate>100</update_rate>

            <!-- Number of wheel pairs -->
            <num_wheel_pairs>2</num_wheel_pairs>

            <!-- wheels0 -->
            <left_joint>front_left_wheel_joint</left_joint>
            <right_joint>front_right_wheel_joint</right_joint>

            <!-- wheels1-->
            <left_joint>back_left_wheel_joint</left_joint>
            <right_joint>back_right_wheel_joint</right_joint>

            <!-- kinematics -->
            <wheel_separation>0.51</wheel_separation>
            <wheel_separation>0.51</wheel_separation>

            <wheel_diameter>0.2</wheel_diameter>
            <wheel_diameter>0.2</wheel_diameter>

            <!-- limits -->
            <max_wheel_torque>50</max_wheel_torque>
            <max_wheel_acceleration>1.0</max_wheel_acceleration>

            <!-- output -->
            <publish_odom>true</publish_odom>
            <publish_odom_tf>true</publish_odom_tf>
            <publish_wheel_tf>false</publish_wheel_tf>

            <odometry_frame>odom_demo</odometry_frame>
            <robot_base_frame>mobile_base_link</robot_base_frame>
        </plugin>
    </gazebo>

    <ros2_control name="GazeboSystem" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>

        <joint name="arm_base_joint">
            <command_interface name="position">
                <param name="min">-${pi}</param>
                <param name="max">${pi}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="link_1_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="link_2_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="link_3_joint">
            <command_interface name="position">
                <param name="min">-${pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="gripper_base_joint">
            <command_interface name="position">
                <param name="min">-${pi}</param>
                <param name="max">${pi}</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        
        <joint name="gripper_finger_left_joint">
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <joint name="gripper_finger_right_joint">
            <param name="mimic">gripper_finger_left_joint</param>
            <param name="multiplier">1</param>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
    </ros2_control>
    
  
  
    <gazebo>
        <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
            <robot_sim_type>gazebo_ros2_control/GazeboSystem</robot_sim_type>
            <parameters>$(find custom_robot_sim)/config/joint_trajectory_p_controller.yaml</parameters>
        </plugin>
    </gazebo>

        <!-- <plugin name="gazebo_ros_joint_pose_trajectory" filename="libgazebo_ros_joint_pose_trajectory.so">
            <ros>
                <namespace>/demo</namespace>
                <remapping>set_joint_trajectory:=set_trajectory_demo</remapping>
            </ros>
            <update_rate>100</update_rate>
        </plugin> -->

</robot>